pkg/cli/debug_merge_logs.go

```diff
@@ -0,0 +1,556 @@
+			return err
+		}
+		return ei.Format(w)
+	}
+
+	g, ctx := errgroup.WithContext(context.Background())
+	entryChan := make(chan entryInfo, chanSize) // read -> bufferWrites
+	writeChan := make(chan *bytes.Buffer)       // bufferWrites -> write
+	read := func() error {
+		defer close(entryChan)
+		for e, ok := s.peek(); ok; e, ok = s.peek() {
+			select {
+			case entryChan <- entryInfo{Entry: e, FileInfo: s.fileInfo()}:
+			case <-ctx.Done():
+				return nil
+			}
+			s.pop()
+		}
+		return s.error()
+	}
+	bufferWrites := func() error {
+		defer close(writeChan)
+		writing, pending := &bytes.Buffer{}, &bytes.Buffer{}
+		for {
+			send, recv := writeChan, entryChan
+			if pending.Len() == 0 {
+				send = nil
+				if recv == nil {
+					return nil
+				}
+			} else if pending.Len() > maxWriteBufSize {
+				recv = nil
+			}
+			select {
+			case ei, open := <-recv:
+				if !open {
+					entryChan = nil
+					break
+				}
+				startLen := pending.Len()
+				if err := render(ei, pending); err != nil {
+					return err
+				}
+				if filter != nil && !filter.Match(pending.Bytes()[startLen:]) {
+					pending.Truncate(startLen)
+				}
+			case send <- pending:
+				writing.Reset()
+				pending, writing = writing, pending
+			case <-ctx.Done():
+				return nil
+			}
+		}
+	}
+	write := func() error {
+		for buf := range writeChan {
+			if _, err := out.Write(buf.Bytes()); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	g.Go(read)
+	g.Go(bufferWrites)
+	g.Go(write)
+	return g.Wait()
+}
+
+// mergedStream is a merged heap of log streams.
+type mergedStream []logStream
+
+// newMergedStreamFromPatterns creates a new logStream by first glob
+// expanding pattern, then filtering for matching files which conform to the
+// log filename pattern and match the program. The returned stream will only
+// return log entries in [from, to].
+func newMergedStreamFromPatterns(
+	ctx context.Context, patterns []string, program *regexp.Regexp, from, to time.Time,
+) (logStream, error) {
+	paths, err := expandPatterns(patterns)
+	if err != nil {
+		return nil, err
+	}
+	files, err := findLogFiles(paths, program, to)
+	if err != nil {
+		return nil, err
+	}
+	return newMergedStream(ctx, files, from, to)
+}
+
+func newMergedStream(
+	ctx context.Context, files []fileInfo, from, to time.Time,
+) (*mergedStream, error) {
+	// TODO(ajwerner): think about clock movement and PID
+	const maxConcurrentFiles = 256 // Should far less than the FD limit.
+	sem := make(chan struct{}, maxConcurrentFiles)
+	res := make(mergedStream, len(files))
+	var g errgroup.Group
+	createFileStream := func(i int) func() error {
+		return func() error {
+			sem <- struct{}{}
+			defer func() { <-sem }()
+			s, err := newFileLogStream(files[i], from, to)
+			if s != nil {
+				res[i] = s
+			}
+			return err
+		}
+	}
+	for i := range files {
+		g.Go(createFileStream(i))
+	}
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+	filtered := res[:0] // Filter the nil streams.
```

Comment:
inconsequential nit, but thought you'd generally like to know: Go comment notation uses proper punctuation when the comment is on its own line, but not in inline comments, so this would technically be `// filter nil streams`. I don't know that I love this rule, though. Happy to ignore it.